### ISSUE TITLE
@GrailsCompileStatic fails on domain class with inner enum & mapping block

### ISSUE THREAD

**Issue by darxriggs** [grails/grails-core#12101]
This issue is similar to #10808 but instead of the `constraints` block it is about the `mapping` block.

### Steps to Reproduce
1. Create a **domain class** that has an **inner enum** and a **mapping block** (which can even be empty).
```
@GrailsCompileStatic
class User {
    String username
    State state

    static mapping = {
        state(enumType: "string", defaultValue: State.ACTIVE)
    }

    enum State { ACTIVE, INACTIVE, UNKNOWN }
}
```

2. Build the project with `./gradlew build`.

### Expected Behaviour

Compilation succeeds.

### Actual Behaviour

Compilation fails.

Error:
```
> Task :compileGroovy
startup failed:
General error during canonicalization: Cannot get property 'initialExpression' on null object

java.lang.NullPointerException: Cannot get property 'initialExpression' on null object
	at org.codehaus.groovy.runtime.NullObject.getProperty(NullObject.java:60)
	at org.codehaus.groovy.runtime.InvokerHelper.getProperty(InvokerHelper.java:190)
	at org.codehaus.groovy.runtime.callsite.NullCallSite.getProperty(NullCallSite.java:46)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGetProperty(AbstractCallSite.java:298)
	at org.grails.compiler.DomainMappingTypeCheckingExtension$_run_closure4.doCall(DomainMappingTypeCheckingExtension.groovy:54)
	...
```

### Environment Information

- **Operating System**: Linux
- **Grails Version:** 4.0.x
- **JDK Version:** 8

### Example Application

https://github.com/jamesdh/gorm-static-inner-enum-bug (already provided in https://github.com/grails/grails-core/issues/10808#issuecomment-655582363 by @jamesdh).

**Comment by cleydsonjr** [comment]
@puneetbehl How can I proceed to pull this fix into 3.3.15?

**Comment by puneetbehl** [comment]
Please update to Grails 4 as Grails 3 is EOL.
